## Laboratory work 1. Data preprocessing.
### Task1
### Work with Titanic dataset with this laboratory work.Now you need to replace missing data for all remaining features. The main goal is to clean data, look at the data, get some information to make some hypothesis.

#### Look at the data, summary statistics and general information, then find rows with missing data for example, Fare using isnull() method. You can observe only one row. So, let's fill it with median. Find median value and as condition mention Pclass and Embarked features. Then fill the missing data with found median, don't forget about inplace argument. Finally show result with info() method.

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [3]:
data = pd.read_csv("titanic-data.csv")

In [4]:
data

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [5]:
data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


### 1. Fill missing data for Embarked feature. 

In [7]:
mean_of_age = int(data["Age"].mean())
data["Age"].fillna(mean_of_age, inplace = True)

In [8]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          891 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


### 2. Next, you have to handle missing data of Age feature. First of all show missing data for Age using isnull() method

### 3. Pay attention to the Name feature, where except name, there is title/position. So, based on this feature we can determine missing data for Age feature.
### First of all, show Name feature. So, in GetTitle(name) method you have to split the name till comma, then again split till dot, next convert the title to lower case and finally return this value of the title.

### 4. Use the method GetTitle() for each name in the dataframe.

### 5. Now, for previous task use unique() method to get unique values of titles from the dataframe.

### 6.Some of the values mean the same, like 'Lady' and 'Dona, it is recommended you to join such kind of values into one. As a result change the method GetTitle(),which will extract the title from the name, however, that time here will be only shorcut values of titles. Use given dict 
                {'mr' : 'Mr', 
               'mrs' : 'Mrs', 
               'miss' : 'Miss', 
               'master' : 'Master',
               'don' : 'Sir',
               'rev' : 'Sir',
               'dr' : 'Officer',
               'mme' : 'Mrs',
               'ms' : 'Mrs',
               'major' : 'Officer',
               'lady' : 'Lady',
               'sir' : 'Sir',
               'mlle' : 'Miss',
               'col' : 'Officer',
               'capt' : 'Officer',
               'the countess' : 'Lady',
               'jonkheer' : 'Sir',
               'dona' : 'Lady'
                 }

### 7. Here, create a new feature, called 'Title'. Use the changed method GetTitle() for each name in the dataframe.

### 8. Create a boxplot, that grouped by Age and Title features and use only not null values of Age feature. Make conclusion based on this boxplot. 

### 9. Create a variable title_age_median and initialize it by grouping data by Title and get Age, then call median() method on self producing a DataFrame with transformed values using transform() method. 

### 10. Fill missing data for Age feature with title_age_median, don't forget about inplace argument.

### 11. Working with outliers. Build a histogram for Age feature and use this histogram to get understand the distribution. What can you say about distribution?

### 12. Show passengers, who are over 70 years old.

### 13. Create a histogram for fare feature, set title 'Histogram for fare', set bins as 20 and set color. Make conclusion based on this histogram.

### 14. Create a boxplot for Fare feature to indentify outliers, what can you say based on this boxplot?

### 15. To look into the outliers show the data, where Fare is maximum.

### 16. Next, let's try some transformations to reduce the skewness. Using the logarithm, let's convert the asymmetric data to more symmetric data. And notice to add 1 to accomodate zero fares : log(0) is not defined. Then show a histogram for the logarithm Fare.

### 17. Using qcut() function, quantile-based discretization function, discretize Fare variable into 4 equal-sized buckets. Produce a Categorical object indicating quantile membership for each data point like 'very_low', 'low', 'high', 'very_high' labels.

### 18. Based on previous task, get value counts and build a barplot. 

### 19. Based on previous task, create a new feature, called Fare_Bin.

### 20. Using info() function, show information about dataframe.

### 21. Feature engineering. 
### Let's divide the passengers into adults and children. Create a new feature, called 'AgeState', using where() function, which takes first argument as condition that Age is greater or equal than 18. If given condition is True, value of new feature is 'Adult', else it is 'Child'.

### 22. Show the new feature AgeState using value_counts() function.

### 23. Using crosstabs() function create a cross tabulation for Survived and AgeState features. Don't forget to exclude -888 value.

### 24. Let's create another feature, called 'FamilySize' by adding Parents with Siblings + 1 for self.
#### Parch - number of parent / children aboard
#### SibSp - number of siblings aboard

### 25. Create a histogram for new FamilySize feature. Make conclusion based on this histogram.

### 26. To look into the outliers show the data, where FamilySize is maximum.

### 27. Using crosstab() function create a cross tabulation for Survived and FamilySize features. Don't forget to exclude -888 value.

### 28. Let's create another feature, called 'isMother', to find this value you need to use where() function, which takes some conditions as arguments. First condition would be that Sex is female, then Parch number is greater than 0, next Age is greater than 18 and suppose that title is not Miss.

### 29. Count the values of isMother feature.

### 30. Using crosstab() function create a cross tabulation for Survived and isMother features. Don't forget to exclude -888 value.

### 31. Next, let's work with Cabin feature. Show it.

### 32. Show unique values of Cabin feature.

### 33. Each value has a letter in upper case and some numbers, whereas you have to notice there is a one value that has no numbers. Suppose, it is an error value. Let's convert it into Nan. 

### 34. Create a function, called get_deck, which takes argument Cabin and returns changed values as 'Z' for NaN values of Cabin argument, and drops numbers of existed values of Cabin. As a result, function returns letter in upper case.

### 35. Let's create another feature, called 'Deck', to find this value you need to call get_deck() function.

### 36. Let's look at dataframe, call info() function, now we have new 5 features.

### 37. Using crosstab() function create a cross tabulation for Survived and Deck features. Don't forget to exclude -888 value.

### 38. Categorical Feature Encoding. Let's work with Gender feature. Create a new feature, called 'isMale', using where() function.

### 39. Using get_dummies() function encode 'Deck', 'Pclas', 'Title', 'Fare_Bin', 'Embarked', 'AgeState' features. Here must be 2 arguments, first is dataframe, second is list of features.

### 40. Call info() function, after encoding we have 34 features, and suppose we don't need 'Cabin', 'Name', 'Ticket', 'Parch', 'SibSp', 'Sex' features. As a result delete them and don't forget about inplace argument.